docs/jax-101/06-parallelism.md

```diff
@@ -0,0 +1,423 @@
+  """Computes the least squares error of the model's predictions on x against y."""
+  pred = params.weight * xs + params.bias
+  return jnp.mean((pred - ys) ** 2)
+
+LEARNING_RATE = 0.005
+
+# So far, the code is identical to the single-device case. Here's what's new:
+
+
+# Remember that the `axis_name` is just an arbitrary string label used
+# to later tell `jax.lax.pmean` which axis to reduce over. Here, we call it
+# 'num_devices', but could have used anything, so long as `pmean` used the same.
+@functools.partial(jax.pmap, axis_name='num_devices')
+def update(params: Params, xs: jnp.ndarray, ys: jnp.ndarray) -> Params:
+  """Performs one SGD update step on params using the given data."""
+
+  # Compute the gradients on the given minibatch (individually on each device).
+  loss, grads = jax.value_and_grad(loss_fn)(params, xs, ys)
+
+  # Combine the gradient across all devices (by taking their mean).
+  grads = jax.lax.pmean(grads, axis_name='num_devices')
+
+  # Also combine the loss. Unnecessary for the update, but useful for logging.
+  loss = jax.lax.pmean(loss, axis_name='num_devices')
+
+  # Each device performs its own update, but since we start with the same params
+  # and synchronise gradients, the params stay in sync.
+  new_params = jax.tree_multimap(
+      lambda param, g: param - g * LEARNING_RATE, params, grads)
+
+  return new_params, loss
+```
+
++++ {"id": "RWce8YZ4Pcmf"}
+
+Here's how `update()` works:
+
+Undecorated and without the `pmean`s, `update()` takes data tensors of shape `[batch, ...]`, computes the loss function on that batch and evaluates its gradients.
+
+We want to spread the `batch` dimension across all available devices. To do that, we add a new axis using `pmap`. The arguments to the decorated `update()` thus need to have shape `[num_devices, batch_per_device, ...]`. So, to call the new `update()`, we'll need to reshape data batches so that what used to be `batch` is reshaped to `[num_devices, batch_per_device]`. That's what `split()` does below. Additionally, we'll need to replicate our model parameters, adding the `num_devices` axis. This reshaping is how a pmapped function knows which devices to send which data.
+
+At some point during the update step, we need to combine the gradients computed by each device -- otherwise, the updates performed by each device would be different. That's why we use `jax.lax.pmean` to compute the mean across the `num_devices` axis, giving us the average gradient of the batch. That average gradient is what we use to compute the update.
+
+Aside on naming: here, we use `num_devices` for the `axis_name` for didactic clarity while introducing `jax.pmap`. However, in some sense that is tautologous: any axis introduced by a pmap will represent a number of devices. Therefore, it's common to see the axis be named something semantically meaningful, like `batch`, `data` (signifying data parallelism) or `model` (signifying model parallelism).
```

Comment:
Another dumb question, sorry: is this meant to be `A side note on naming` or `Aside from naming`. (I see the PR is `pull ready` but if it's not too late).